source/server/listener_manager_impl.cc

```diff
@@ -94,16 +95,77 @@ ProdListenerComponentFactory::createListenSocket(Network::Address::InstanceConst
   if (fd != -1) {
     ENVOY_LOG(debug, "obtained socket for address {} from parent", addr);
-    return std::make_shared<Network::TcpListenSocket>(fd, address);
+    return std::make_shared<Network::TcpListenSocket>(fd, address, options);
   }
-  return std::make_shared<Network::TcpListenSocket>(address, bind_to_port);
+  return std::make_shared<Network::TcpListenSocket>(address, options, bind_to_port);
 }
 
 DrainManagerPtr
 ProdListenerComponentFactory::createDrainManager(envoy::api::v2::Listener::DrainType drain_type) {
   return DrainManagerPtr{new DrainManagerImpl(server_, drain_type)};
 }
 
+// Socket::Option implementation for API-defined listener socket options.
+// This same object can be extended to handle additional listener socket options.
+class ListenerSocketOption : public Network::Socket::Option, Logger::Loggable<Logger::Id::config> {
+public:
+  ListenerSocketOption(const envoy::api::v2::Listener& config)
+      : transparent_(config.transparent()) {}
+
+  // Network::Socket::Option
+  bool setOption(Network::Socket& socket, Network::Socket::SocketState state) const override {
+    // IP_TRANSPARENT has an effect on bind() (allowing bind() to non-local addresses), but
+    // also after bind(), controlling whether new connections with non-local addresses are
+    // redirected to the listening socket or not, so we need to set it in either case, as
+    // it is possible to, e.g., set it on a listener and then change the listener configuration
+    // to not set it.
+    // This has the side effect of setting the option twice after the socket is first created,
+    // both before and after bind().
+    UNREFERENCED_PARAMETER(state);
+
+    const auto* ip = socket.localAddress()->ip();
+    if (!ip) {
+      return false;
+    }
+#ifdef SOL_IP
+    int rc;
+    int transparent = transparent_;
+
+    if (ip->version() == Network::Address::IpVersion::v4) {
+      rc = setsockopt(socket.fd(), SOL_IP, IP_TRANSPARENT, &transparent, sizeof(transparent));
+    } else {
+#if defined(SOL_IPV6) && defined(IPV6_TRANSPARENT)
+      rc = setsockopt(socket.fd(), SOL_IPV6, IPV6_TRANSPARENT, &transparent, sizeof(transparent));
+#else
+      rc = -1;
+      errno = ENOTSUP;
+#endif
+    }
+    // If we are unable to set the option we still return success for the default value (0),
+    // as we never changed the option to begin with. This also prevents failures in all current
+    // test cases that do not set the transparent option.
+    if (rc == 0 || !transparent_) {
+      return true;
+    }
+#else
+    UNREFERENCED_PARAMETER(socket);
+
+    if (!transparent_) {
+      return true;
+    }
+    errno = ENOTSUPP;
+#endif
+    ENVOY_LOG(warn, "Setting IP_TRANSPARENT to {} on listener socket failed: {}", transparent,
+              strerror(errno));
+    // Throw a specific message we can verify in testing.
+    throw EnvoyException("ListenSocketOption: Error setting transparent socket option");
```

Comment:
You're throwing from a function that is documented to return false on failure. Either update the docs and make it return void, or make this not throw.

Reply:
Removed the throw and adjusted testing accordingly.